py4web/core.py

```diff
@@ -872,44 +865,47 @@ def decorator(func):
 
             if Fixture.__fixture_debug__:
                 # in debug mode log all calls to fixtures
-                def call(obj, f, args=()):
-                    logging.debug("Calling %s.%s", obj.__class__.__name__, f)
-                    return getattr(obj, f)(*args)
+                def call(f, context):
+                    logging.debug(
+                        f"Calling {f.__self__.__class__.__name__}.{f.__name__}"
+                    )
+                    return f(context)
 
             else:
 
-                def call(obj, f, args=()):
-                    return getattr(obj, f)(*args)
+                def call(f, context):
+                    return f(context)
 
             @functools.wraps(func)
             def wrapper(*args, **kwargs):
                 # data shared by all fixtures in the pipeline for each request
-                shared_data = {"template_context": {}}
+                processed = []
+                context = {
+                    "fixtures": fixtures,
+                    "status": 200,
+                    "output": None,
+                    "exception": None,
+                    "processed": processed,
+                }
                 try:
-                    for obj in fixtures:
-                        call(obj, "on_request")
-                    ret = func(*args, **kwargs)
-                    for obj in fixtures:
-                        ret = call(obj, "transform", (ret, shared_data))
-                    for obj in fixtures:
-                        call(obj, "on_success", (200,))
-                    return ret
+                    for fixture in fixtures:
+                        call(fixture.on_request, context)
+                        processed.append(fixture)
+                    context["output"] = func(*args, **kwargs)
                 except HTTP as http:
-                    if http.type == http.Type.success:
-                        for obj in fixtures:
-                            call(obj, "on_success", (http.status,))
-                    else:
-                        if obj in fixtures:
-                            call(obj, "on_error")
-                    raise
-                except Exception:
-                    for obj in fixtures:
-                        call(obj, "on_error")
-                    raise
+                    context["status"] = http.status
+                    raise http
+                except Exception as error:
+                    context["exception"] = error
+                    if not isinstance(error, HTTP):
+                        for fixture in reversed(processed):
+                            call(fixture.on_error, context)
```

Comment:
What if an exception goes here (error in fixture)?

Reply:
If there is an error in a fixture this results in a ticket. I do not think it is possible for fixtures to handle errors in other fixtures (except on_request).

Reply:
so if we have an exception in fixture before db-fixture, how does db-fixture release the connection?

Reply:
Good point. Let me fix that.

Reply:
On second thought. What if there is an error in the fixture that is supposed to release the connection?

Reply:
Addressed but I do not think this can be made fool proof.

Reply:
> On second thought. What if there is an error in the fixture that is supposed to release the connection?

I guess It is a bug or total crash. But in case of an exception in before-fixture it could be part of the app-logic